clustering/server/src/main/java/org/wildfly/clustering/server/provider/CacheServiceProviderRegistry.java

```diff
@@ -139,51 +178,57 @@ void register(Node node, T service) {
 
     @Override
     public void membershipChanged(List<Node> previousMembers, List<Node> members, final boolean merged) {
-        this.executor.execute(() -> {
-            if (this.getGroup().isCoordinator()) {
-                Set<Node> deadNodes = new HashSet<>(previousMembers);
-                deadNodes.removeAll(members);
-                Set<Node> newNodes = new HashSet<>(members);
-                newNodes.removeAll(previousMembers);
-                if (!deadNodes.isEmpty()) {
-                    try (Batch batch = this.batcher.createBatch()) {
-                        try (CloseableIterator<Map.Entry<T, Set<Node>>> entries = this.cache.entrySet().iterator()) {
-                            while (entries.hasNext()) {
-                                Map.Entry<T, Set<Node>> entry = entries.next();
-                                Set<Node> nodes = entry.getValue();
-                                if (nodes.removeAll(deadNodes)) {
-                                    entry.setValue(nodes);
-                                }
+        if (this.getGroup().isCoordinator()) {
+            Set<Node> deadNodes = new HashSet<>(previousMembers);
+            deadNodes.removeAll(members);
+            Set<Node> newNodes = new HashSet<>(members);
+            newNodes.removeAll(previousMembers);
+            if (!deadNodes.isEmpty()) {
+                try (Batch batch = this.batcher.createBatch()) {
+                    try (CloseableIterator<Map.Entry<T, Set<Node>>> entries = this.cache.entrySet().iterator()) {
+                        while (entries.hasNext()) {
+                            Map.Entry<T, Set<Node>> entry = entries.next();
+                            Set<Node> nodes = entry.getValue();
+                            if (nodes.removeAll(deadNodes)) {
+                                entry.setValue(nodes);
                             }
                         }
                     }
                 }
-                if (merged) {
-                    try (Batch batch = this.batcher.createBatch()) {
-                        for (Node node: newNodes) {
-                            // Re-assert services for new members following merge since these may have been lost following split
-                            CacheServiceProviderRegistry.this.getServices(node).forEach(service -> this.register(node, service));
-                        }
+            }
+            if (merged) {
```

Comment:
The previous `if (!deadNodes.isEmpty())` logic block is generally mutually exclusive with this `if (merged)` logic block.